### ISSUE TITLE
Encode endianness

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#76]
I propose to encode endianness in format:
```
e([bl]\d+_?)+
```
which are consequentially ```endianness```, ```endianness type``` and ```group size``` (by default 8 bits) and the consequential groups mean endianness within the groups (BE assumed).

The examples for  0x0123456789ABCDEF
```eb32``` or ```eb16```  or ```eb8``` or  ```eb4``` or "eb32b16b8b4b2"
```0123456789ABCDEF```

```el32``` or ```el32b32``` or ```el32b8```
```89ABCDEF 01234567```

```el32b16l8```
```AB/89|EF/CD 23/01|67/45```

```el16```
```CDEF 89AB 4567 0123```

```el8```
```EF CD AB 89 67 45 23 01```

```el4```
```F E D C B A 9 8 7 6 5 4 3 2 1 0```

**Comment by GreyCat** [comment]
I don't understand it as well. What do you mean by "encoding endianness"?

**Comment by KOLANICH** [comment]
>I don't fully understand where should we use this form.

We should be able to use that form everywhere where we use ```be``` and ```le``` now.

>I don't understand it as well. What do you mean by "encoding endianness"?

I mean that endianness is not limited to "big endian" and "little endian". There are also "mixed endian" architectures. I can imagine even more endianneses though I've never seen them IRL. So I propose a way to encode a large set of endianneses for the case if someone needed them. See the examples I have provided.

**Comment by koczkatamas** [comment]
Interesting proposal. I've never seen this kind of mixed-endianness in any of the file formats I meet till now.

Do you know any file format which uses this kind of endianness?

The problem looks like to me as a special encoding, eg. uleb128. Supporting these kind of encodings are possible now by using instances in a bit hacky way.

So I would handle this issue as a low priority until we figure out how mainstream the usage of mixed-encodings..

**Comment by skochinsky** [comment]
One possible example of mixed endianness is ARM BE-8 where data is big-endian but instructions use little-endian encodings. Also for ARM, the old Maverick FPA used [mixed-endian format  (45670123) for doubles](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0056d/Bcfhgcgd.html). However, I'm not sure if this is something that should be handled on the ksy level.

**Comment by GreyCat** [comment]
As far as I understand, "data is big-endian but instructions use little-endian encodings" is *not* what's called "mixed endianness". Doubles mixed endian format looks closer to what's discussed.

**Comment by davidhicks** [comment]
Mixed endian support is required for ESRI Shapefiles where different fields in the same structure can be big endian or little endian. For reference, see the specification: https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf

**Comment by KOLANICH** [comment]
Here is an [impl in python](https://gitlab.com/KOLANICH/Endianness.py), though`simplify` func is not very optimal.

**Comment by fdwr** [comment]
>old Maverick FPA used mixed-endian format (45670123) for doubles

@skochinsky: Yeah, I've been collecting a list. Of course 8-bit endianness reversals are the most common, but I've found plenty of 16-bit endianness reversals too. There's really no such thing as "middle endianness", just endianness of different unit sizes (e.g. 8 bits swapped per every 32 bits, 16 bits swapped per 32 bits, 36 bits swapped per 72...), and it becomes clear that "middle endianness" is just a case of big endianness using a larger word size than 8-bits, when looking at examples like the 128-bit VAX H_floating point format. The VAX formats are all reversed 16-bit chunks (https://nssdc.gsfc.nasa.gov/nssdc/formats/VAXFloatingPoint.htm), along with some Honeywell ones, and PDP-11. The most novel is the Univac 72-bit format which swaps the two 36-bit subwords, not that any of those file formats matter much anymore (good luck even finding tape drives outside a museum). Depending on how the data was serialized to disk, the 6-bit bytes can be zero padded to 8-bits, instead yielding an endianness of 48-bits per 96.

```
int32be16:{endian:{16 32} bits:32 integer:31 sign:1 notes:"PDP-11 https://en.wikipedia.org/wiki/Endianness#PDP-endian, http://bitsavers.org/pdf/dec/pdp11/handbooks/PDP1145_Handbook_1973.pdf page165"}
int32be16Honeywell:{bits:32 integer:15..30 sign:1 integer:0..15 unused:1 /*the unused "positive bit" is always zero*/ notes:"Honeywell 316 float32 CDAB order, with most significant 16 bits first and a weird unused gap bit between the 16-bit words https://en.wikipedia.org/wiki/Endianness#Honeywell_316 http://archive.computerhistory.org/resources/text/Honeywell/Honeywell.H316.1965.102646157.pdf https://archive.org/details/bitsavers_honeywells156316516PgmrRefNov70_4279319/page/n17/mode/2up"}
float32f23e8bias128s1be16:{endian:{16 32} bits:32 fraction:23 hiddenOne:true exponent:{bits:8 bias:128} sign:1 encoding:signAndMagnitude notes:"DEC PDP-11 float32 CDAB order https://en.wikipedia.org/wiki/Endianness#PDP-endian, http://bitsavers.org/pdf/dec/pdp11/handbooks/PDP1145_Handbook_1973.pdf page165 7.4. VAX-11 F_floating format https://nssdc.gsfc.nasa.gov/nssdc/formats/VAXFloatingPoint.htm"}
float48f23e8s1be16:{endian:{16 48} bits:48 fraction:39 hiddenOne:true exponent:{bits:8 bias:127 /*?*/} sign:1 encoding:signAndMagnitude notes:"Honeywell 316 float32 EFCDAB order, with most significant 16 bits first https://en.wikipedia.org/wiki/Endianness#Honeywell_316 http://archive.computerhistory.org/resources/text/Honeywell/Honeywell.H316.1965.102646157.pdf page 8 https://archive.org/details/bitsavers_honeywells156316516PgmrRefNov70_4279319/page/n17/mode/2up *uncertain exponent bias"}
float64f52e11bias1024s1be16:{endian:{16 64} bits:64 fraction:52 hiddenOne:true exponent:{bits:11 bias:1024} sign:1 encoding:signAndMagnitude notes:"VAX-11 G_floating format https://nssdc.gsfc.nasa.gov/nssdc/formats/VAXFloatingPoint.htm"}
float64f55e8bias128s1be16:{endian:{16 64} bits:64 fraction:55 hiddenOne:true exponent:{bits:8 bias:128} sign:1 encoding:signAndMagnitude notes:"DEC PDP-11 float64 GHEFCDAB order https://en.wikipedia.org/wiki/Endianness#PDP-endian, http://bitsavers.org/pdf/dec/pdp11/handbooks/PDP1145_Handbook_1973.pdf page165. VAX-11 D_floating format https://nssdc.gsfc.nasa.gov/nssdc/formats/VAXFloatingPoint.htm"}
float72f60e11s1be36:{endian:{36 72} bits:72 fraction:60 hiddenOne:false exponent:{bits:11 bias:1024} sign:1 encoding:signAndMagnitude notes:"UNIVAC 72 bits. Has no hidden bit. Least significant fraction bits are in second 36-bit word. https://en.wikipedia.org/wiki/UNIVAC_1100/2200_series, https://en.wikipedia.org/wiki/Floating-point_arithmetic#History, https://nssdc.gsfc.nasa.gov/nssdc/formats/UnivacUnisys36bit.htm"}
float128f112e15bias16384s1be16:{endian:{16 128} bits:128 fraction:112 hiddenOne:true exponent:{bits:15 bias:16384} sign:1 encoding:signAndMagnitude notes:"VAX-11 H_floating format https://nssdc.gsfc.nasa.gov/nssdc/formats/VAXFloatingPoint.htm"}
```
